bindings/serialize.cc

```diff
@@ -0,0 +1,397 @@
+                                         allocation->size,
+                                         p.getStringId("bytes"))};
+      std::vector<int64_t> vals = {
+          allocation->count,
+          static_cast<int64_t>(allocation->size * allocation->count)};
+      Sample s = Sample({stack.begin(), stack.end()}, vals, labels);
+      samples.push_back(s);
+    }
+    return samples;
+  }
+};
+
+class TimeNode : public Node {
+ private:
+  const CpuProfileNode *node;
+  int samplingIntervalMicros;
+
+ public:
+  TimeNode(const CpuProfileNode *node, int samplingIntervalMicros) {
+    this->node = node;
+    this->samplingIntervalMicros = samplingIntervalMicros;
+  }
+  virtual std::string getName() {
+    return *String::Utf8Value(node->GetFunctionName());
+  }
+  virtual std::string getFilename() {
+    return *String::Utf8Value(node->GetScriptResourceName());
+  }
+  virtual int64_t getFileID() { return node->GetScriptId(); }
+  virtual int64_t getLineNumber() { return node->GetLineNumber(); }
+  virtual int64_t getColumnNumber() { return node->GetColumnNumber(); }
+  virtual std::vector<Sample> getSamples(Profile &p,
+                                         std::deque<uint64_t> stack) {
+    std::vector<Sample> samples;
+    int64_t hitCount = node->GetHitCount();
+    std::vector<int64_t> vals = {hitCount, hitCount * samplingIntervalMicros};
+    Sample s = Sample({stack.begin(), stack.end()}, vals, std::vector<Label>());
+    samples.push_back(s);
+    return samples;
+  }
+};
+
+struct TimeEntry {
+  const CpuProfileNode *node;
+  std::deque<uint64_t> stack;
+};
+
+std::vector<char> serializeTimeProfile(CpuProfile *profile,
+                                       int64_t samplingIntervalMicros,
+                                       int64_t startTimeNanos) {
+  int64_t durationNanos =
+      (profile->GetEndTime() - profile->GetStartTime()) * 1000;
+
+  Profile p = Profile("wall", "microseconds", samplingIntervalMicros,
+                      startTimeNanos, durationNanos);
+  p.addSampleType("sample", "count");
+  p.addSampleType("wall", "microseconds");
+
+  // Add root to entries
+  const CpuProfileNode *root = profile->GetTopDownRoot();
+  std::vector<TimeEntry> entries;
+  int32_t count = root->GetChildrenCount();
+  for (int32_t i = 0; i < count; i++) {
+    TimeEntry child;
+    child.node = root->GetChild(i);
+    entries.push_back(child);
+  }
+
+  // Iterate over profile tree and serialize
+  while (entries.size() > 0) {
+    TimeEntry entry = entries.back();
+    entries.pop_back();
+
+    std::unique_ptr<Node> node(
+        new TimeNode(entry.node, samplingIntervalMicros));
+    uint64_t loc = p.addSample(node, entry.stack);
+    std::deque<uint64_t> stack = entry.stack;
+    stack.push_front(loc);
+    int32_t count = entry.node->GetChildrenCount();
+    for (int32_t i = 0; i < count; i++) {
+      TimeEntry child;
+      child.node = entry.node->GetChild(i);
+      child.stack = stack;
```

Comment:
I don't think we have to copy the stack over and over again. This code can maintain the current stack as it walks over the tree and then add samples passing that.

Reply:
Done.